File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/components/hardware/DriveTrain.java
package org.firstinspires.ftc.teamcode.components.hardware;

import static java.lang.Math.sqrt;

import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorEx;
import com.qualcomm.robotcore.hardware.DcMotorSimple;
import com.qualcomm.robotcore.hardware.Gamepad;
import com.qualcomm.robotcore.hardware.HardwareMap;

import org.firstinspires.ftc.robotcore.external.Telemetry;
import org.firstinspires.ftc.teamcode.components.device.GamepadKt;
import org.firstinspires.ftc.teamcode.components.meta.Hardware;
import org.firstinspires.ftc.teamcode.components.meta.MotorGroup;

public class DriveTrain {
    HardwareMap hardwareMap;
    Telemetry telemetry;

    DcMotorEx frontLeft, frontRight, backLeft, backRight;

    MotorGroup motorGroup;
    public DriveTrain(HardwareMap hardwareMap, Telemetry telemetry) {
        this.hardwareMap = hardwareMap;
        this.telemetry = telemetry;
        frontLeft = hardwareMap.get(DcMotorEx.class, Hardware.DRIVE_FL);// Motor vars
        frontLeft.setDirection(DcMotorSimple.Direction.REVERSE);
        frontRight = hardwareMap.get(DcMotorEx.class, Hardware.DRIVE_FR);
        backLeft = hardwareMap.get(DcMotorEx.class, Hardware.DRIVE_BL);
        backLeft.setDirection(DcMotorSimple.Direction.REVERSE);
        backRight = hardwareMap.get(DcMotorEx.class, Hardware.DRIVE_BR);
        motorGroup = new MotorGroup(frontLeft, frontRight, backLeft, backRight);

        motorGroup.applyToMotors(motor -> motor.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER));
        motorGroup.applyToMotors(motor -> motor.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE));
    }

    private double largest(double[] arr) {
        double max = Math.abs(arr[0]);
        int i = 1;
        while (i < arr.length) {
            if (Math.abs(arr[i]) > max) {
                max = Math.abs(arr[i]);
            }
            i++;
        }
        return max;
    }

    private double[] reduce(double[] x) {
        double max = largest(x);
        if (max > 1) {
            for (int i = 0; i < x.length; i++) {
                x[i] = x[i] / max;
            }
        }
        return x;
    }


    public void drive(Gamepad gamepad, double powerMulti) {
        double[] driveSticks = GamepadKt.getDriveSticks(gamepad);
        double x = driveSticks[0];
        double y = driveSticks[1];
        double r = driveSticks[2];

        double theta = Math.atan2(y, x);
        double power = Math.min(Math.hypot(x, y), 1.0);

        double xComponent = power * Math.cos(theta - Math.PI / 4);
        double yComponent = power * Math.sin(theta - Math.PI / 4);

        double[] powers = {
                xComponent + r,
                yComponent - r,
                yComponent + r,
                xComponent - r
        };


//        if (power + Math.abs(r) > 1) {
//            for (int i = 0; i < powers.length; i++) {
//                powers[i] /= (power + Math.abs(r));
//            }
//        }

        double _powerMulti = !GamepadKt.isAnyJoystickTriggered(gamepad) ? 0.0 : powerMulti;

        for (int i = 0; i < powers.length; i++) {
            powers[i] = Math.pow(powers[i], 3.0) * _powerMulti;
        }

        powers = reduce(powers);

        double[] finalPowers = powers;
        motorGroup.applyToMotors(motor -> motor.setPower(finalPowers[motorGroup.motorIndex]));
        telemetry.addData("FL", frontLeft.getPower());
        telemetry.addData("FR", frontRight.getPower());
        telemetry.addData("BL", backLeft.getPower());
        telemetry.addData("BR", backRight.getPower());
        telemetry.update();

    }


}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/components/hardware/DriveTrain.kt
@file:Suppress("LocalVariableName")
@file:Config

package org.firstinspires.ftc.teamcode.components.hardware

/* ?
? * Drivetrain Component
! * Don't call this directly if using BotComponents
? *
 * TODO: N/A
? */

import com.acmerobotics.dashboard.config.Config
import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.DcMotorSimple.Direction
import com.qualcomm.robotcore.hardware.Gamepad
import com.qualcomm.robotcore.hardware.HardwareMap
//import ftc.rogue.blacksmith.util.kt.pow
import org.firstinspires.ftc.teamcode.components.device.getDriveSticks
import org.firstinspires.ftc.teamcode.components.device.isAnyJoystickTriggered
import org.firstinspires.ftc.teamcode.components.meta.DeviceNames
import org.firstinspires.ftc.teamcode.components.utils.maxMagnitudeAbs
import org.firstinspires.ftc.teamcode.components.utils.pow
import kotlin.math.*

@JvmField
var tipCorrection = true
@JvmField
var tipMult = 7.0
@JvmField
var tipDeadzone = 0.05

class DriveTrain (hardwareMap: HardwareMap) {
    private val frontLeft  = hardwareMap.get(DcMotorEx::class.java, DeviceNames.DRIVE_FL).apply { direction = Direction.REVERSE } // Motor vars
    private val frontRight = hardwareMap.get(DcMotorEx::class.java, DeviceNames.DRIVE_FR).apply { direction = Direction.REVERSE } // ! added a flip cuz of anderson powerpole being flipped
    private val backLeft   = hardwareMap.get(DcMotorEx::class.java, DeviceNames.DRIVE_BL).apply { direction = Direction.REVERSE }
    private val backRight  = hardwareMap.get(DcMotorEx::class.java, DeviceNames.DRIVE_BR)

    init {
        withEachMotor {
            zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
            mode = DcMotor.RunMode.RUN_WITHOUT_ENCODER
        }
    }

    fun drive(gamepad: Gamepad, powerMulti: Double) { // ? drive function
        val (x, y, _r) = gamepad.getDriveSticks() // look at com.qualcomm.robotcore.hardware.Gamepad

        val r = _r * 1.0f // * doing the math
        val theta = atan2(y, x)
        val power = hypot(x, y) // sqrt(x.pow(2) + y.pow(2)).coerceAtMost(1F)

        var xComponent = power * cos(theta - PI / 4) // * more math
        var yComponent = power * sin(theta - PI / 4)

        // * unused tip correction function
//        if(tipCorrection){
//             Strafe correction
//            if(abs(b) > tipDeadzone) {
//                xComponent += tipMult * b
//                yComponent -= tipMult * b
//            }
//        }
//

        val max = maxMagnitudeAbs<Double>(xComponent, yComponent, 1e-16) // truly one of the most legible functions of all time, peek at package org.firstinspires.ftc.teamcode.components.utils.MathUtils if you wanna suffer

        val powers = doubleArrayOf( // wow power array
            power * (xComponent / max) + r,
            power * (yComponent / max) - r,
            power * (yComponent / max) + r,
            power * (xComponent / max) - r,
        )

        if (power + abs(r) > 1) { // reduce under 1 for power
            powers.mapInPlace { it / (power + abs(r)) }
        }

        val _powerMulti = if (!gamepad.isAnyJoystickTriggered()) 0.0 else powerMulti

        for (i in powers.indices) { // cube all powers for more precise manuevering
            powers[i] = powers[i].pow(3.0) * _powerMulti
        }

        withEachMotor {
            this.power = powers[it]
        }
    }

    private fun DoubleArray.mapInPlace(transform: (Double) -> Double) = repeat(size) {
        this[it] = transform(this[it])
    }

    private fun withEachMotor(transformation: DcMotorEx.(Int) -> Unit) {
        frontLeft .transformation(0)
        frontRight.transformation(1)
        backLeft  .transformation(2)
        backRight .transformation(3)
    }
}